packages/backend/src/apps/m365-excel/FOR_RELEASE_PERIOD_ONLY.ts

```diff
@@ -0,0 +1,55 @@
+import type { IGlobalVariable } from '@plumber/types'
+
+import { RateLimiterRedis, RateLimiterRes } from 'rate-limiter-flexible'
+
+import { M365_EXCEL_INTERVAL_BETWEEN_ACTIONS_MS } from '@/config/app-env-vars/m365'
+import { createRedisClient, REDIS_DB_INDEX } from '@/config/redis'
+import StepError from '@/errors/step'
+
+const redisClient = createRedisClient(REDIS_DB_INDEX.RATE_LIMIT)
+
+// Rate limiter to control potential thundering herd during release priod
+// --
+// This allows a burst from test steps up to 2x our allocated M365 quota.
+//
+// Instead of rate limiting per second, we allow spikes in 15 second windows to
+// make the UX not _as_ bad.
+//
+// (e.g. thundering herd spams tests within the 1st 10 seconds and use up all
+// quota allocated for 15s, but they don't complain since they're busy
+// publishing or thinking about why they had an error)
+const ACTIONS_PER_SECOND = 1000 / M365_EXCEL_INTERVAL_BETWEEN_ACTIONS_MS
+const DURATION_SECONDS = 30
+const ACTIONS_PER_DURATION = Math.floor(ACTIONS_PER_SECOND * DURATION_SECONDS)
+
+const rateLimiter = new RateLimiterRedis({
+  points: ACTIONS_PER_DURATION,
+  duration: DURATION_SECONDS,
+  keyPrefix: 'm365-release-limiter',
+  storeClient: redisClient,
+})
```

Comment:
hmm if the `M365_EXCEL_INTERVAL_BETWEEN_ACTIONS_MS` is 1000, then is this going to be 1 execution step per second for all pipes (over a period of 30s), just tryna sanity check

Reply:
30 per 30 seconds, but if it consumes 30 in the first second, then the remaining 29s will be rate-limited